crates/bevy_ecs/src/index/mod.rs

```diff
@@ -0,0 +1,456 @@
+                storage_type,
+                Layout::new::<()>(),
+                None,
+                false,
+            )
+        };
+
+        world.register_component_with_descriptor(descriptor)
+    }
+
+    /// Gets the [`ComponentId`]s of all markers that _must_ be included on an [`Entity`] allocated
+    /// to a particular slot index.
+    fn ids_for(&self, index: usize) -> Vec<ComponentId> {
+        self.markers
+            .iter()
+            .enumerate()
+            .filter_map(|(i, &id)| (index & (1 << i) > 0).then_some(id))
+            .collect::<Vec<_>>()
+    }
+}
+
+/// Options when configuring an index for a given indexable component `C`.
+pub struct IndexOptions<
+    C: Component<Mutability = Immutable>,
+    S: IndexStorage<C> = HashMap<C, usize>,
+> {
+    /// Marker components will be added to indexed entities to allow for efficient lookups.
+    /// This controls the [`StorageType`] that will be used with these markers.
+    /// It is recommended to use [`Table`](StorageType::Table) for more efficient querying, but a
+    /// [`SpareSet`](StorageType::SparseSet) may be more appropriate in certain circumstances.
+    ///
+    /// This defaults to [`Table`](StorageType::Table).
+    pub marker_storage: StorageType,
+    /// Marker components are combined into a unique address for each distinct value of the indexed
+    /// component.
+    /// This controls how many markers will be used to create that unique address.
+    /// Note that a value greater than 32 will be reduced down to 32.
+    /// Bevy's [`World`] only supports 2^32 entities alive at any one moment in time, so an address space
+    /// of 32 is sufficient to uniquely refer to every individual [`Entity`] in the entire [`World`].
+    ///
+    /// Selecting a non-default value may lead to a panic at runtime or entities missing from the
+    /// index if the address space is exhausted.
+    ///
+    /// This defaults to [`size_of<C>`]
+    pub address_space: usize,
+    /// A storage backend for this index.
+    /// For certain indexing strategies and [`Component`] types, you may be able to greatly
+    /// optimize the utility and performance of an index by creating a custom backend.
+    ///
+    /// See [`IndexStorage`] for details around the implementation of a custom backend.
+    ///
+    /// This defaults to [`HashMap<C, usize>`].
+    pub index_storage: S,
+    #[doc(hidden)]
+    pub _phantom: PhantomData<fn(&C)>,
+}
+
+impl<C: Component<Mutability = Immutable> + Eq + Hash + Clone> Default
+    for IndexOptions<C, HashMap<C, usize>>
+{
+    fn default() -> Self {
+        Self {
+            marker_storage: StorageType::Table,
+            address_space: size_of::<C>(),
+            index_storage: HashMap::with_hasher(FixedHasher),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+/// Extension methods for [`World`] to assist with indexing components.
```

Comment:
We're living in bevy_ecs already; just make these methods on World. See above about wanting to only have a single method though.